File: scripts/validate_toe.py
#!/usr/bin/env python3
"""
TouchDesigner ASCII Validator

Validates expanded .toe.dir projects before collapse.
Catches errors that TD silently fails on.

Usage:
    python validate_toe.py ./project.toe.dir
    python validate_toe.py ./project.toe.dir --verbose
"""

import os
import sys
import ast
import re
from pathlib import Path
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class ValidationResult:
    errors: list = field(default_factory=list)
    warnings: list = field(default_factory=list)
    info: list = field(default_factory=list)

    def add_error(self, file: str, line: Optional[int], msg: str):
        loc = f"{file}:{line}" if line else file
        self.errors.append(f"{loc} - {msg}")

    def add_warning(self, file: str, msg: str):
        self.warnings.append(f"{file} - {msg}")

    def add_info(self, msg: str):
        self.info.append(msg)

    @property
    def has_errors(self) -> bool:
        return len(self.errors) > 0

    @property
    def has_warnings(self) -> bool:
        return len(self.warnings) > 0


VALID_FAMILIES = {"TOP", "CHOP", "SOP", "DAT", "COMP", "MAT", "POP"}


def validate_n_file(filepath: Path, result: ValidationResult) -> Optional[dict]:
    """Validate a .n node definition file. Returns node info if valid."""

    rel_path = filepath.name
    try:
        content = filepath.read_text()
    except Exception as e:
        result.add_error(rel_path, None, f"Cannot read file: {e}")
        return None

    lines = content.strip().split('\n')

    if not lines:
        result.add_error(rel_path, None, "Empty file")
        return None

    # Check first line: FAMILY:type
    first_line = lines[0].strip()
    if ':' not in first_line:
        result.add_error(rel_path, 1, f"Invalid format, expected FAMILY:type, got: {first_line}")
        return None

    family, op_type = first_line.split(':', 1)
    if family not in VALID_FAMILIES:
        result.add_error(rel_path, 1, f"Invalid family '{family}', must be one of: {VALID_FAMILIES}")
        return None

    # Check for 'end' keyword
    if not lines[-1].strip() == 'end':
        result.add_error(rel_path, len(lines), "File must end with 'end'")

    # Check for required sections
    has_tile = any(line.strip().startswith('tile ') for line in lines)
    has_flags = any(line.strip().startswith('flags =') for line in lines)
    has_color = any(line.strip().startswith('color ') for line in lines)

    if not has_tile:
        result.add_warning(rel_path, "Missing 'tile' definition")
    if not has_flags:
        result.add_warning(rel_path, "Missing 'flags' definition")
    if not has_color:
        result.add_warning(rel_path, "Missing 'color' definition")

    # Extract inputs (wiring)
    inputs = {}
    in_inputs_block = False
    for i, line in enumerate(lines, 1):
        stripped = line.strip()

        if stripped == 'inputs':
            in_inputs_block = True
            continue

        if in_inputs_block:
            if stripped == '{':
                continue
            if stripped == '}':
                in_inputs_block = False
                continue

            # Check for TAB vs spaces
            if '\t' not in line and stripped:
                result.add_error(rel_path, i, f"inputs entry uses SPACES, must use TAB: '{stripped}'")
            else:
                # Parse input: INDEX\tSOURCE_NAME
                parts = stripped.split('\t')
                if len(parts) >= 2:
                    try:
                        idx = int(parts[0])
                        source = parts[1].strip()
                        inputs[idx] = source
                    except ValueError:
                        result.add_error(rel_path, i, f"Invalid input index: {parts[0]}")

    node_name = filepath.stem
    return {
        'name': node_name,
        'family': family,
        'type': op_type,
        'inputs': inputs,
        'path': str(filepath)
    }


def validate_parm_file(filepath: Path, result: ValidationResult):
    """Validate a .parm parameter file."""

    rel_path = filepath.name
    try:
        content = filepath.read_text()
    except Exception as e:
        result.add_error(rel_path, None, f"Cannot read file: {e}")
        return

    lines = content.strip().split('\n')

    if not lines:
        result.add_warning(rel_path, "Empty parameter file")
        return

    # Check for ? delimiters
    if lines[0].strip() != '?':
        result.add_error(rel_path, 1, "Parameter file must start with '?'")

    if lines[-1].strip() != '?':
        result.add_error(rel_path, len(lines), "Parameter file must end with '?'")

    # Check parameter format
    for i, line in enumerate(lines[1:-1], 2):
        stripped = line.strip()
        if not stripped:
            continue

        parts = stripped.split(None, 2)  # Split into max 3 parts
        if len(parts) < 2:
            result.add_warning(rel_path, f"Line {i}: Invalid format, expected 'name flags value'")
            continue

        # Check flags
        try:
            flags = int(parts[1])
            if flags not in (0, 17):
                result.add_warning(rel_path, f"Line {i}: Unusual flags value {flags} (typically 0 or 17)")
        except ValueError:
            result.add_error(rel_path, i, f"Invalid flags value: {parts[1]}")


def validate_text_file(filepath: Path, result: ValidationResult):
    """Validate a .text script file (binary format)."""

    rel_path = filepath.name
    try:
        content = filepath.read_bytes()
    except Exception as e:
        result.add_error(rel_path, None, f"Cannot read file: {e}")
        return

    # Check binary header
    if len(content) < 27:
        result.add_error(rel_path, None, f"File too short ({len(content)} bytes), missing binary header (need 27+ bytes)")
        return

    # Check format marker
    if not content.startswith(b'2\n*'):
        result.add_error(rel_path, None, "Missing binary header - file appears to be plain text, not TD format")
        result.add_info(f"  Hint: Use write_td_text_file() from the editing guide to create proper .text files")
        return

    # Extract Python content (skip 27-byte header)
    try:
        script_content = content[27:].decode('utf-8')
    except UnicodeDecodeError as e:
        result.add_error(rel_path, None, f"Cannot decode script content: {e}")
        return

    # Validate Python syntax
    try:
        ast.parse(script_content)
    except SyntaxError as e:
        result.add_error(rel_path, e.lineno, f"Python SyntaxError: {e.msg}")

    # Check for ChopExec callbacks
    if 'chopexec' in filepath.stem.lower() or 'chop_exec' in filepath.stem.lower():
        expected_callbacks = ['onOffToOn', 'onOnToOff', 'onValueChange', 'whileOn']
        for cb in expected_callbacks:
            if cb not in script_content:
                result.add_warning(rel_path, f"ChopExec missing callback: {cb}")


def validate_toc(toc_path: Path, dir_path: Path, result: ValidationResult):
    """Validate .toc against actual files."""

    try:
        toc_content = toc_path.read_text()
    except Exception as e:
        result.add_error(".toc", None, f"Cannot read file: {e}")
        return

    toc_entries = set(line.strip() for line in toc_content.split('\n') if line.strip())

    # Check all toc entries exist
    for entry in toc_entries:
        entry_path = dir_path / entry
        if not entry_path.exists():
            result.add_error(".toc", None, f"Listed file does not exist: {entry}")

    # Check all files are in toc (excluding hidden files)
    for file_path in dir_path.rglob('*'):
        if file_path.is_file():
            rel = file_path.relative_to(dir_path)
            rel_str = str(rel)

            # Skip hidden files and toc itself
            if rel_str.startswith('.') or rel_str == '.toc':
                continue

            if rel_str not in toc_entries:
                result.add_warning(f".toc", f"File not listed in .toc: {rel_str}")


def validate_wiring(nodes: dict, result: ValidationResult):
    """Validate all wiring references exist."""

    node_names = set(nodes.keys())

    for name, node in nodes.items():
        for idx, source in node.get('inputs', {}).items():
            if source not in node_names:
                # Check for near matches
                close_matches = [n for n in node_names if n.lower() == source.lower()]
                if close_matches:
                    result.add_error(
                        node['path'], None,
                        f"Input {idx} references '{source}' (not found). Did you mean '{close_matches[0]}'?"
                    )
                else:
                    result.add_error(
                        node['path'], None,
                        f"Input {idx} references non-existent node: '{source}'"
                    )

    # Check for orphans (no incoming connections)
    referenced = set()
    for node in nodes.values():
        for source in node.get('inputs', {}).values():
            referenced.add(source)

    for name in node_names:
        if name not in referenced:
            # Only warn if it's not a typical source (noise, constant, etc.)
            node = nodes[name]
            if node['type'] not in ('noise', 'constant', 'moviefilein', 'audiodevicein', 'midiin', 'oscin'):
                result.add_warning(node['path'], f"No incoming connections (orphan source)")


def validate_project(dir_path: Path, verbose: bool = False) -> ValidationResult:
    """Validate an expanded .toe.dir project."""

    result = ValidationResult()

    if not dir_path.exists():
        result.add_error(str(dir_path), None, "Directory does not exist")
        return result

    if not dir_path.is_dir():
        result.add_error(str(dir_path), None, "Path is not a directory")
        return result

    # Collect all nodes
    nodes = {}
    n_count = 0
    parm_count = 0
    text_count = 0

    # Validate .n files
    for n_file in dir_path.rglob('*.n'):
        n_count += 1
        node_info = validate_n_file(n_file, result)
        if node_info:
            nodes[node_info['name']] = node_info

    # Validate .parm files
    for parm_file in dir_path.rglob('*.parm'):
        parm_count += 1
        validate_parm_file(parm_file, result)

    # Validate .text files
    for text_file in dir_path.rglob('*.text'):
        text_count += 1
        validate_text_file(text_file, result)

    # Validate .toc
    toc_path = dir_path / '.toc'
    if toc_path.exists():
        validate_toc(toc_path, dir_path, result)
    else:
        # Try parent directory for .toc
        parent_toc = dir_path.parent / f"{dir_path.name.replace('.dir', '.toc')}"
        if parent_toc.exists():
            validate_toc(parent_toc, dir_path, result)
        else:
            result.add_warning(".toc", "No .toc file found")

    # Validate wiring
    validate_wiring(nodes, result)

    # Summary
    result.add_info(f"Checked: {n_count} .n files, {parm_count} .parm files, {text_count} .text files")

    return result


def main():
    if len(sys.argv) < 2:
        print("Usage: python validate_toe.py <project.toe.dir> [--verbose]")
        print("\nValidates expanded TouchDesigner projects before collapse.")
        sys.exit(2)

    dir_path = Path(sys.argv[1])
    verbose = '--verbose' in sys.argv or '-v' in sys.argv

    print(f"Validating: {dir_path}\n")

    result = validate_project(dir_path, verbose)

    # Print results
    for info in result.info:
        print(f"  {info}")

    print()

    if result.errors:
        print(f"ERRORS ({len(result.errors)}):")
        for err in result.errors:
            print(f"  ✗ {err}")
        print()

    if result.warnings:
        print(f"WARNINGS ({len(result.warnings)}):")
        for warn in result.warnings:
            print(f"  ⚠ {warn}")
        print()

    if not result.errors and not result.warnings:
        print("✓ All checks passed!")
    elif result.errors:
        print(f"✗ {len(result.errors)} error(s), {len(result.warnings)} warning(s)")
    else:
        print(f"⚠ {len(result.warnings)} warning(s), no errors")

    # Exit code
    if result.errors:
        sys.exit(2)
    elif result.warnings:
        sys.exit(1)
    else:
        sys.exit(0)


if __name__ == '__main__':
    main()

File: scripts/validate-toe.ts
#!/usr/bin/env bun
/**
 * TouchDesigner ASCII Validator (TypeScript)
 *
 * Validates expanded .toe.dir projects before collapse.
 * Catches errors that TD silently fails on.
 *
 * Usage:
 *   bun scripts/validate-toe.ts ./project.toe.dir
 *   bun scripts/validate-toe.ts ./project.toe.dir --verbose
 */

import {
  readFileSync,
  readdirSync,
  existsSync,
  statSync,
} from "fs";
import { join, basename, extname } from "path";

// ─── Types ─────────────────────────────────────────────────────────

export interface NodeInfo {
  name: string;
  family: string;
  type: string;
  inputs: Record<number, string>;
  path: string;
}

// ─── ValidationResult ──────────────────────────────────────────────

export class ValidationResult {
  errors: string[] = [];
  warnings: string[] = [];
  info: string[] = [];

  addError(file: string, line: number | null, msg: string): void {
    const loc = line != null ? `${file}:${line}` : file;
    this.errors.push(`${loc} - ${msg}`);
  }

  addWarning(file: string, msg: string): void {
    this.warnings.push(`${file} - ${msg}`);
  }

  addInfo(msg: string): void {
    this.info.push(msg);
  }

  get hasErrors(): boolean {
    return this.errors.length > 0;
  }

  get hasWarnings(): boolean {
    return this.warnings.length > 0;
  }
}

// ─── Constants ─────────────────────────────────────────────────────

const VALID_FAMILIES = new Set([
  "TOP",
  "CHOP",
  "SOP",
  "DAT",
  "COMP",
  "MAT",
  "POP",
]);

// Types that are expected to have no incoming references (sources and terminals)
const ORPHAN_EXEMPT_TYPES = new Set([
  "noise",
  "constant",
  "moviefilein",
  "audiodevicein",
  "midiin",
  "oscin",
  "null",
]);

// ─── Helpers ───────────────────────────────────────────────────────

function walkDir(dir: string): string[] {
  const results: string[] = [];
  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...walkDir(full));
    } else if (entry.isFile()) {
      results.push(full);
    }
  }
  return results;
}

// ─── Validators ────────────────────────────────────────────────────

export function validateNFile(
  filepath: string,
  result: ValidationResult
): NodeInfo | null {
  const relPath = basename(filepath);
  let content: string;
  try {
    content = readFileSync(filepath, "utf-8");
  } catch (e: any) {
    result.addError(relPath, null, `Cannot read file: ${e.message}`);
    return null;
  }

  const lines = content.trim().split("\n");

  if (lines.length === 0 || (lines.length === 1 && lines[0].trim() === "")) {
    result.addError(relPath, null, "Empty file");
    return null;
  }

  // Check first line: FAMILY:type
  const firstLine = lines[0].trim();
  if (!firstLine.includes(":")) {
    result.addError(
      relPath,
      1,
      `Invalid format, expected FAMILY:type, got: ${firstLine}`
    );
    return null;
  }

  const colonIdx = firstLine.indexOf(":");
  const family = firstLine.substring(0, colonIdx);
  const opType = firstLine.substring(colonIdx + 1);

  if (!VALID_FAMILIES.has(family)) {
    result.addError(
      relPath,
      1,
      `Invalid family '${family}', must be one of: ${[...VALID_FAMILIES].join(", ")}`
    );
    return null;
  }

  // Check for 'end' keyword
  if (lines[lines.length - 1].trim() !== "end") {
    result.addError(relPath, lines.length, "File must end with 'end'");
  }

  // Check for required sections
  const hasTile = lines.some((l) => l.trim().startsWith("tile "));
  const hasFlags = lines.some((l) => l.trim().startsWith("flags ="));
  const hasColor = lines.some((l) => l.trim().startsWith("color "));

  if (!hasTile) result.addWarning(relPath, "Missing 'tile' definition");
  if (!hasFlags) result.addWarning(relPath, "Missing 'flags' definition");
  if (!hasColor) result.addWarning(relPath, "Missing 'color' definition");

  // Extract inputs (wiring)
  const inputs: Record<number, string> = {};
  let inInputsBlock = false;

  for (let i = 0; i < lines.length; i++) {
    const stripped = lines[i].trim();
    const lineNum = i + 1;

    if (stripped === "inputs") {
      inInputsBlock = true;
      continue;
    }

    if (inInputsBlock) {
      if (stripped === "{") continue;
      if (stripped === "}") {
        inInputsBlock = false;
        continue;
      }

      // Check for TAB vs spaces
      if (!lines[i].includes("\t") && stripped) {
        result.addError(
          relPath,
          lineNum,
          `inputs entry uses SPACES, must use TAB: '${stripped}'`
        );
      } else {
        // Parse input: INDEX\tSOURCE_NAME
        const parts = stripped.split("\t");
        if (parts.length >= 2) {
          const idx = parseInt(parts[0], 10);
          if (isNaN(idx)) {
            result.addError(
              relPath,
              lineNum,
              `Invalid input index: ${parts[0]}`
            );
          } else {
            const source = parts[1].trim();
            inputs[idx] = source;
          }
        }
      }
    }
  }

  const nodeName = basename(filepath, extname(filepath));
  return {
    name: nodeName,
    family,
    type: opType,
    inputs,
    path: relPath,
  };
}

export function validateParmFile(
  filepath: string,
  result: ValidationResult
): void {
  const relPath = basename(filepath);
  let content: string;
  try {
    content = readFileSync(filepath, "utf-8");
  } catch (e: any) {
    result.addError(relPath, null, `Cannot read file: ${e.message}`);
    return;
  }

  const lines = content.trim().split("\n");

  if (lines.length === 0) {
    result.addWarning(relPath, "Empty parameter file");
    return;
  }

  // Check for ? delimiters
  if (lines[0].trim() !== "?") {
    result.addError(relPath, 1, "Parameter file must start with '?'");
  }

  if (lines[lines.length - 1].trim() !== "?") {
    result.addError(
      relPath,
      lines.length,
      "Parameter file must end with '?'"
    );
  }

  // Check parameter format (lines between delimiters)
  for (let i = 1; i < lines.length - 1; i++) {
    const stripped = lines[i].trim();
    if (!stripped) continue;

    const parts = stripped.split(/\s+/);
    if (parts.length < 2) {
      result.addWarning(
        relPath,
        `Line ${i + 1}: Invalid format, expected 'name flags value'`
      );
      continue;
    }

    // Check flags
    const flags = parseInt(parts[1], 10);
    if (isNaN(flags)) {
      result.addError(relPath, i + 1, `Invalid flags value: ${parts[1]}`);
    } else if (flags !== 0 && flags !== 17) {
      result.addWarning(
        relPath,
        `Line ${i + 1}: Unusual flags value ${flags} (typically 0 or 17)`
      );
    }
  }
}

export function validateTextFile(
  filepath: string,
  result: ValidationResult
): void {
  const relPath = basename(filepath);
  let buf: Buffer;
  try {
    buf = readFileSync(filepath) as Buffer;
  } catch (e: any) {
    result.addError(relPath, null, `Cannot read file: ${e.message}`);
    return;
  }

  // Check format marker first: 2\n*  (0x32, 0x0A, 0x2A)
  if (buf.length < 3 || buf[0] !== 0x32 || buf[1] !== 0x0a || buf[2] !== 0x2a) {
    result.addError(
      relPath,
      null,
      "Missing binary header - file appears to be plain text, not TD format"
    );
    result.addInfo(
      `  Hint: Use write_td_text_file() from the editing guide to create proper .text files`
    );
    return;
  }

  // Check minimum size for valid binary .text file
  if (buf.length < 27) {
    result.addError(
      relPath,
      null,
      `File too short (${buf.length} bytes), missing binary header (need 27+ bytes)`
    );
    return;
  }

  // Extract script content (skip 27-byte header)
  let scriptContent: string;
  try {
    scriptContent = buf.subarray(27).toString("utf-8");
  } catch (e: any) {
    result.addError(relPath, null, `Cannot decode script content: ${e.message}`);
    return;
  }

  // Check for ChopExec callbacks
  const stem = basename(filepath, extname(filepath)).toLowerCase();
  if (stem.includes("chopexec") || stem.includes("chop_exec")) {
    const expectedCallbacks = [
      "onOffToOn",
      "onOnToOff",
      "onValueChange",
      "whileOn",
    ];
    for (const cb of expectedCallbacks) {
      if (!scriptContent.includes(cb)) {
        result.addWarning(relPath, `ChopExec missing callback: ${cb}`);
      }
    }
  }
}

export function validateToc(
  tocPath: string,
  dirPath: string,
  result: ValidationResult
): void {
  let tocContent: string;
  try {
    tocContent = readFileSync(tocPath, "utf-8");
  } catch (e: any) {
    result.addError(".toc", null, `Cannot read file: ${e.message}`);
    return;
  }

  const tocEntries = new Set(
    tocContent
      .split("\n")
      .map((l) => l.trim())
      .filter((l) => l.length > 0)
  );

  // Check all toc entries exist as files
  for (const entry of tocEntries) {
    const entryPath = join(dirPath, entry);
    if (!existsSync(entryPath)) {
      result.addError(".toc", null, `Listed file does not exist: ${entry}`);
    }
  }

  // Check all files in dirPath are listed in toc (skip hidden files)
  const allFiles = walkDir(dirPath);
  for (const filePath of allFiles) {
    const rel = filePath.substring(dirPath.length + 1); // strip dirPath + separator
    if (rel.startsWith(".")) continue;
    if (!tocEntries.has(rel)) {
      result.addWarning(".toc", `File not listed in .toc: ${rel}`);
    }
  }
}

export function validateWiring(
  nodes: Record<string, NodeInfo>,
  result: ValidationResult
): void {
  const nodeNames = new Set(Object.keys(nodes));

  // Check all input references exist
  for (const [, node] of Object.entries(nodes)) {
    for (const [idx, source] of Object.entries(node.inputs)) {
      if (!nodeNames.has(source)) {
        // Check for case-insensitive near matches
        const closeMatches = [...nodeNames].filter(
          (n) => n.toLowerCase() === source.toLowerCase()
        );
        if (closeMatches.length > 0) {
          result.addError(
            node.path,
            null,
            `Input ${idx} references '${source}' (not found). Did you mean '${closeMatches[0]}'?`
          );
        } else {
          result.addError(
            node.path,
            null,
            `Input ${idx} references non-existent node: '${source}'`
          );
        }
      }
    }
  }

  // Check for orphans (no incoming connections)
  const referenced = new Set<string>();
  for (const node of Object.values(nodes)) {
    for (const source of Object.values(node.inputs)) {
      referenced.add(source);
    }
  }

  for (const name of nodeNames) {
    if (!referenced.has(name)) {
      const node = nodes[name];
      if (!ORPHAN_EXEMPT_TYPES.has(node.type)) {
        result.addWarning(
          node.path,
          "No incoming connections (orphan source)"
        );
      }
    }
  }
}

export function validateProject(
  dirPath: string,
  verbose: boolean = false
): ValidationResult {
  const result = new ValidationResult();

  if (!existsSync(dirPath)) {
    result.addError(dirPath, null, "Directory does not exist");
    return result;
  }

  const stat = statSync(dirPath);
  if (!stat.isDirectory()) {
    result.addError(dirPath, null, "Path is not a directory");
    return result;
  }

  // Collect all files
  const allFiles = walkDir(dirPath);
  const nodes: Record<string, NodeInfo> = {};
  let nCount = 0;
  let parmCount = 0;
  let textCount = 0;

  // Validate .n files
  for (const f of allFiles) {
    if (extname(f) === ".n") {
      nCount++;
      const nodeInfo = validateNFile(f, result);
      if (nodeInfo) {
        nodes[nodeInfo.name] = nodeInfo;
      }
    }
  }

  // Validate .parm files
  for (const f of allFiles) {
    if (extname(f) === ".parm") {
      parmCount++;
      validateParmFile(f, result);
    }
  }

  // Validate .text files
  for (const f of allFiles) {
    if (extname(f) === ".text") {
      textCount++;
      validateTextFile(f, result);
    }
  }

  // Validate .toc
  const tocInDir = join(dirPath, ".toc");
  if (existsSync(tocInDir)) {
    validateToc(tocInDir, dirPath, result);
  } else {
    // Try parent directory for .toc
    const dirName = basename(dirPath);
    const parentToc = join(
      dirPath,
      "..",
      dirName.replace(".dir", ".toc")
    );
    if (existsSync(parentToc)) {
      validateToc(parentToc, dirPath, result);
    } else {
      result.addWarning(".toc", "No .toc file found");
    }
  }

  // Validate wiring
  validateWiring(nodes, result);

  // Summary
  result.addInfo(
    `Checked: ${nCount} .n files, ${parmCount} .parm files, ${textCount} .text files`
  );

  return result;
}

// ─── CLI ───────────────────────────────────────────────────────────

function main(): void {
  const args = process.argv.slice(2);

  if (args.length === 0 || args.every((a) => a.startsWith("-"))) {
    console.log(
      "Usage: bun scripts/validate-toe.ts <project.toe.dir> [--verbose]"
    );
    console.log(
      "\nValidates expanded TouchDesigner projects before collapse."
    );
    process.exit(2);
  }

  const dirPath = args.find((a) => !a.startsWith("-"))!;
  const verbose = args.includes("--verbose") || args.includes("-v");

  console.log(`Validating: ${dirPath}\n`);

  const result = validateProject(dirPath, verbose);

  // Print results
  for (const info of result.info) {
    console.log(`  ${info}`);
  }

  console.log();

  if (result.errors.length > 0) {
    console.log(`ERRORS (${result.errors.length}):`);
    for (const err of result.errors) {
      console.log(`  \u2717 ${err}`);
    }
    console.log();
  }

  if (result.warnings.length > 0) {
    console.log(`WARNINGS (${result.warnings.length}):`);
    for (const warn of result.warnings) {
      console.log(`  \u26A0 ${warn}`);
    }
    console.log();
  }

  if (!result.hasErrors && !result.hasWarnings) {
    console.log("\u2713 All checks passed!");
  } else if (result.hasErrors) {
    console.log(
      `\u2717 ${result.errors.length} error(s), ${result.warnings.length} warning(s)`
    );
  } else {
    console.log(
      `\u26A0 ${result.warnings.length} warning(s), no errors`
    );
  }

  // Exit code
  if (result.hasErrors) {
    process.exit(2);
  } else if (result.hasWarnings) {
    process.exit(1);
  } else {
    process.exit(0);
  }
}

// Run CLI if executed directly
if (import.meta.main) {
  main();
}
